### PR TITLE
Rule S3984: Exception should not be created without being thrown

### LINKED ISSUE #251: Rule S3984: Exception should not be created without being thrown
Implements [RSPEC-3984](https://jira.sonarsource.com/browse/RSPEC-3984)

### PR DESCRIPTION
**Pull request by valhristov**
Fix #251

### CODE REVIEW

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/ExceptionNotThrown.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        }
+
+        public Exception Test2()
+        {
+            return new Exception(); // Compliant
+        }
+
+        public Exception Test3()
+        {
+            var e = new Exception(); // Noncompliant, why not directly return?
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/ExceptionNotThrown.cs]: Some people prefer this style, IMHO this should produce the same result as Test2.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/ExceptionNotThrown.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
+using System;
+using System.Linq;
+using System.Collections.Generic;
+
+namespace Tests.Diagnostics
+{
+    class ExceptionNotThrown
+    {
+        public void Test1()
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/ExceptionNotThrown.cs]: A bit more descriptive test names would be more useful. Not a big deal in this case.

**Comment by valhristov** [comment]
FPs removed!

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ExceptionNotThrown.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class ExceptionNotThrown : SonarDiagnosticAnalyzer
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ExceptionNotThrown.cs]: I am not a big fan of the class name but the only thing I can think of is `ExceptionShouldBeUsed`/`ExceptionShouldBeThrown`

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/ExceptionNotThrown.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    class ExceptionNotThrown
+    {
+        void NotCompliant()
+        {
+            new Exception(); // Noncompliant {{Throw this exception or remove this useless statement.}}
+        }
+
+        Exception Compliant()
+        {
+            var a = new Exception();
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/ExceptionNotThrown.cs]: Everyone suggested that the dead store rule will take care of this

**Pull request merged**